trajectory_tracker/src/trajectory_tracker.cpp

```diff
@@ -401,55 +303,27 @@ void TrackerNode::control()
     return;
   }
 
-  float min_dist_ = 10000.0;
-  int iclose = -1;
-  geometry_msgs::Point origin;
-  origin.x = cos(w_ * look_forward_ / 2.0) * v_ * look_forward_;
-  origin.y = sin(w_ * look_forward_ / 2.0) * v_ * look_forward_;
+  const Eigen::Vector2d origin =
+      Eigen::Vector2d(std::cos(w_lim_.get()), std::sin(w_lim_.get())) *
+      (look_forward_ / 2.0) * v_lim_.get() * look_forward_;
+
+  const double path_length = lpath.length();
+
   // Find nearest line strip
-  out_of_line_strip_ = false;
-  float distance_path_ = 0;
-  for (size_t i = 1; i < lpath.poses.size(); i++)
-  {
-    distance_path_ += dist2d(lpath.poses[i - 1].pose.position, lpath.poses[i].pose.position);
-  }
-  float distance_path_search_ = 0;
-  float sign_vel_prev_ = 0;
-  for (size_t i = path_step_done_; i < lpath.poses.size(); i++)
-  {
-    if (i < 1)
-      continue;
-    distance_path_search_ += dist2d(lpath.poses[i - 1].pose.position, lpath.poses[i].pose.position);
-    if (dist2d(origin, lpath.poses[i].pose.position) < 0.05 &&
-        i < lpath.poses.size() - 1)
-      continue;
-    float d = dist2d_linestrip(lpath.poses[i - 1].pose.position, lpath.poses[i].pose.position, origin);
-    if (fabs(d) <= fabs(min_dist_))
-    {
-      min_dist_ = d;
-      iclose = i;
-    }
-    if (path_step_done_ > 0 && distance_path_search_ > 1.0)
-      break;
-
-    geometry_msgs::Point vec = sub2d(lpath.poses[i].pose.position,
-                                     lpath.poses[i - 1].pose.position);
-    float angle = atan2(vec.y, vec.x);
-    float angle_pose_;
-    if (allow_backward_)
-      angle_pose_ = tf2::getYaw(lpath.poses[i].pose.orientation);
-    else
-      angle_pose_ = angle;
-    float signVel_req = cos(angle) * cos(angle_pose_) + sin(angle) * sin(angle_pose_);
-    if (sign_vel_prev_ * signVel_req < 0)
-    {
-      // Stop read forward if the path_ switched back
-      break;
-    }
-    sign_vel_prev_ = signVel_req;
-  }
-  if (iclose < 0)
+  const trajectory_tracker::Path2D::ConstIterator it_local_goal =
+      lpath.findLocalGoal(
+          lpath.begin(), lpath.end(), allow_backward_);
+
+  const float max_search_range = (path_step_done_ > 0) ? 1.0 : 0.0;
+  const trajectory_tracker::Path2D::ConstIterator it_nearest =
+      lpath.findNearest(
+          lpath.begin() + std::max(1, path_step_done_), it_local_goal, origin,
+          max_search_range);
+
+  if (it_nearest == lpath.end())
```

Comment:
TrajectoryTrackerTest/CurveFollow sometimes entering here

Reply:
in this failure state of the test rviz always dead with
```
rviz: /build/ogre-1.9-mqY1wq/ogre-1.9-1.9.0+dfsg1/OgreMain/include/OgreAxisAlignedBox.h:252: void Ogre::AxisAlignedBox::setExtents(const Ogre::Vector3&, const Ogre::Vector3&): Assertion `(min.x <= max.x && min.y <= max.y && min.z <= max.z) && "The minimum corner of the box must be less than or equal to maximum corner"' failed.
Aborted (core dumped)
```
maybe the data is broken